### PR TITLE
Race condition fixes using the shutdown() trick, timeout handling speedup, slots preallocation

### PR DESCRIPTION
**Pull request by solardiz**
Hi David!

This is the patch I posted to the onion-dev list a while ago.  It isn't exactly clean, but it does the job of making onion work for me seemingly reliably for now (and faster than before, too), and I am not currently aware of any remaining fd or data race conditions in the poller with this patch applied, assuming that the epoll one-shot property works perfectly.  (There may still be issues with fd's getting closed on error by other parts of onion, which would trigger races in the poller again.  You might want to look into that.)

I think the fixtimeouts branch is a dead-end (slow, and possibly a less complete fix), at least for the time being, and this patch should be merged instead.  This way, we can announce that there were those fatal reliability and security bugs when running with a thread pool and epoll, which are now believed to be fixed.  Further cleanups (such as getting rid of the #if 0'ed portions, some of which are actually unsafe, or/and making more invasive changes) can then proceed separately (and will require review and extensive re-testing, as it is all too easy to introduce race conditions again).  Maybe you'll need to make a numbered release after merging this patch and before further related changes, so that there is a released version that is (at least potentially) reliable and secure (unlike what we had before, and unlike what further changes might bring), even if still dirty under the hood.

Thanks,

Alexander

### CODE REVIEW

**Comment by solardiz** [comment]
I just found the timeout vs. event race was reported in mid-2015 as issue #116, still open.  This patch fixes it.

**Comment by davidmoreno** [comment]
Hi,

I can not merge this pull request as it has many `#if 0` and uses a 1M array for connections, which is oversized for normal use (but that might be useful in some cases). 

I will leave it open until I do the merge and testing for the shutdown part, and a numbered release as you suggest.

I will review your points about the fixtimeouts branch, but I think it do fix the stated problems, and it is more robust.

David.

**Comment by solardiz** [comment]
There's no 1M array.  The allocation is based on getrlimit(RLIMIT_NOFILE), which is normally 1024.  The 1M constant is merely the maximum, for the unlikely case the value returned by getrlimit() is even higher than that.

As to the `#if 0`, how about you first merge this pull request as-is, and then remove the `#if 0`'ed portions with a separate commit?  If we check that poller.c.o remains unchanged after a rebuild with those portions removed, then no re-testing is needed.

In what way is the fixtimeouts branch more robust, in your opinion?  It still depends on the one-shot property, right?  Is its dependency on this property working perfectly not as strict, and why?

**Comment by davidmoreno** [comment]
If we use a fixed array for the slots, there is no need for the slot struct to have a `next` pointer, so it should be removed. Also the array should be per poller, not static, as there could be two onion instances running at the same time in the same program. There would be interactions if you set the same fd on two pollers, but it should be possible. So the `onion_poller_slot_new` is unnecesary and `onion_poller_slot_get` could be used instead (API break :( ). I checked if we really need the linked list of current active waiters and I think we can avoid it, if only for the timeouts. Check bellow for a solution.

As I understand it, now it wakes up every second, and do the check for expiring on all slots, where before it calculated how long to wait and then check for expiring slots. Is it right?

I dont like so much that it waits one second and do work, when normally there is just no request being made, so it could wait until new requests. I think we can add a timerfd slot that wakes up on next timeout, do the closing of fds if necessary, and loop over all the slots to check the new expiration time. This expiration time has to be reset as well on new slots and slot function dispatching, as user could set a new timeout sooner than the current timeout. timerfd allows to reset a running timer.

Looping over the fds may look expensive, and benchmarking should be done, but as it should be memory close, as we use an array, it could be even faster than chasing pointers in a linked list. There are cache implications on the array vs linked list, but here in the timeouts we benefit from it as we only read. 

But only for big number of fds, so maybe we can keep a maxfd counter to check as little fds as possible, as I did in my fixtimeouts branch. In sparse situations could be worse, on normal ones it should be faster than the linked list.

Just for the tip of the cake, we have a cache problem if several cores try to access adjacent memory regions in write mode as two subsequent requests with adjacent fds. First of all just now we are even worse, so no need for more improvement here. So maybe we should force a slot size of 64 bytes (which looks like typical cache line size). but indeed some performance testing should be done.

I'm not asking for you to do all this work, but just your thoughts, so we do a proper full fix.

Regards,
David.

**Comment by solardiz** [comment]
> If we use a fixed array for the slots, there is no need for the slot struct to have a next pointer, so it should be removed.

I had this thought too, however for now I wanted to minimize the number of changes relative to your code. Also, as you point out, the list of active fd's can be sparse, and the slots are best put into their own cache lines each. With an array, we'd be hitting some unused cache lines (for closed fd's) while traversing it; with a linked list, we'd hit only the currently active fd's. OTOH, removing a slot from a linked list affects another slot's next pointer, and thus another cache line. So it's not immediately obvious which way is better, and for now we just need to get something reliable enough for a release.

> Also the array should be per poller, not static, as there could be two onion instances running at the same time in the same program. There would be interactions if you set the same fd on two pollers, but it should be possible.

How can it possibly be reasonable to have two pollers share an fd? If the two pollers are configured the same (handlers, etc.), there's no point in having two rather than one. If they're different, then why would an app want to "randomly" handle requests coming over the same fd differently?

> As I understand it, now it wakes up every second, and do the check for expiring on all slots, where before it calculated how long to wait and then check for expiring slots. Is it right?

Yes. Moreover, this happens separately in each thread. It is easy to improve this so that it happens in only one of the threads (IIRC, you already did in fixtimeouts?), but if done the obvious way the other threads would often end up waiting on a mutex, so there would be little performance improvement (for this specific service). I chose to postpone improving this for a later rewrite.

> I dont like so much that it waits one second and do work, when normally there is just no request being made, so it could wait until new requests.

Yes, I don't like this either, but it's the cheaper way of doing things on busy servers.

> I think we can add a timerfd slot that wakes up on next timeout, do the closing of fds if necessary, and loop over all the slots to check the new expiration time. This expiration time has to be reset as well on new slots and slot function dispatching, as user could set a new timeout sooner than the current timeout. timerfd allows to reset a running timer.

This sounds good to me. I am also considering using timerfd's if/when I eventually replace my use of onion with own code.

Waking up on each exact timeout may be more costly than waking up once per second. So you'll want to group nearby timeouts, with some reasonable granularity.

> Just for the tip of the cake, we have a cache problem if several cores try to access adjacent memory regions in write mode as two subsequent requests with adjacent fds. First of all just now we are even worse, so no need for more improvement here. So maybe we should force a slot size of 64 bytes (which looks like typical cache line size). but indeed some performance testing should be done.

Yes, aligning the slots array and padding each slot to cache line size would be good, to avoid false sharing. This would not fully avoid the cache line bouncing, though, since with our design there's also a lot of true sharing: any thread can advance in handling a request that started arriving on another thread (and this property is in fact good for minimizing the network service's latency). Granted, the latter situation might be a write vs. read (rather than write vs. write), but it may also involve transfer of a cache line from another CPU.

I did some related performance testing recently, and determined that writing to a shared cache line on a modern dual-socket system with 8-core CPUs (a total of 32 logical) once per request appears to cost about 0.03% of performance at 10k requests/second (cumulative for 32 threads; or 300 per thread). I simulated this with a busy loop, then scaled down to the 10k request rate (this might or might not be accurate). If we have multiple such accesses per request (probably true for the slots), or/and multiple shared pieces of data like this, or/and if the request rate is much higher, this can add up to a more significant percentage of CPU time.

To summarize: I think all of what you describe here should be revisited after the release.

**Comment by zacharygrafton** [comment]
I'm inclined to agree with @solardiz about revisiting after the 0.8 release. From my testing this patch definitely fixes some issues that show up when compiling with clang and ThreadSanitizer.

**Comment by davidmoreno** [comment]
I just merged the patch almost as id. Only a slight problem with freeing the slots at onion_poller_free.

All the rest of the work is still on the pipeline. I will add issues about it. Specially the wake up every second.

Thanks for the patch.

Regards,
David.

**Comment by solardiz** [comment]
> Only a slight problem with freeing the slots at onion_poller_free.

Oh, indeed.  Sorry about this, and thank you for fixing it.  I missed it because onion_poller_free() is never called in my app, which just runs the poller until the app itself is killed.
